Sources/Utility/Process.swift

```diff
@@ -165,6 +172,24 @@ public final class Process: ObjectIdentifierProtocol {
         self.redirectOutput = redirectOutput 
     }
 
+    /// Validate that the program provided exists on the search path and is executable.
+    internal func validateArgumentIsExecutable(_ program: String) throws {
+
+        // check to see if the argument is already listed in the vaildated cache
+        guard !Process.validatedExecutables.contains(program) else {
+            return
+        }
+        let envSearchPaths = Utility.getEnvSearchPaths(pathString: getenv("PATH"), currentWorkingDirectory: currentWorkingDirectory)
+        guard let executableLocation = Utility.lookupExecutablePath(filename: program, searchPaths: envSearchPaths) else {
+            throw Process.Error.missingProgram(program: program)
+        }
+        guard localFileSystem.isExecutableFile(executableLocation) else {
```

Comment:
Based on my own past experiences of not quite understanding why something failed, I was leaning towards being as absolutely explicit about the error as I could be, presuming that for a CLI, a more detailed error would be useful to someone getting their development or CI environment debugged, and in the case of a library that the additional error types would be more useful. 

That said, the only places currently using this are in UserToolchain and now here in process, so if the specificity doesn't seem really needed, I can collapse it down to a "missing or not executable" pretty easily and shorten even this checking code. I'll update `Utility/Misc.swift` to have `lookupExecutablePath` only return a value if it's both found & executable, which can reduce the number of subsequent guards here and simplify this a bit.